File: 42-trapping-rain-water/trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int sz = height.size();
        int ans = 0;
        vector<int> pre(sz, 0), suf(sz, 0);
        
        for(int i=0; i<sz; i++){
            if(i-1 >= 0){
                pre[i] = max(pre[i-1], height[i-1]);
                suf[sz-1-i] = max(suf[sz-i], height[sz-1-i]);
            }
            else{
                pre[i] = height[i];
                suf[sz-1-i] = height[sz-1-i];
            }
        }

        for(int i=1; i<sz-1; i++){
            if(min(pre[i], suf[i]) > height[i]){
                ans += min(pre[i], suf[i]) - height[i];
            }
        }

        return ans;
    }
};

File: 42-trapping-rain-water/trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int sz = height.length, ans = 0;
        int arr[] = new int[sz], prefix[] = new int[sz], sufix[] = new int[sz];

        prefix[0] = height[0];
        for(int i=1; i<sz; i++){
            if(prefix[i-1] < height[i]) prefix[i] = height[i];
            else prefix[i] = Math.max(prefix[i-1], height[i-1]);
        }

        sufix[sz-1] = height[sz-1];
        for(int i=sz-2; i>=0; i--){
            if(sufix[i+1] < height[i]) sufix[i] = height[i];
            else sufix[i] = Math.max(sufix[i+1], height[i+1]);
        }

        for(int i=0; i<sz; i++){
            // System.out.print(prefix[i] + ":" + sufix[i] + ", ");
            if(Math.min(prefix[i], sufix[i]) > height[i]) ans += Math.min(prefix[i], sufix[i]) - height[i];
        }

        return ans;
    }
}